Runescape chat has some nicer formatting and colours based on jobs.

PowerfulBacon opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Changes some of the formattings for the overhead chat system.

<details><summary>Image showcase</summary>
<p>

**Whispers and radio messages are now small instead of italic**
![image](https://user-images.githubusercontent.com/26465327/80214907-86213a00-8633-11ea-8946-1491cc58d036.png)

**Megaphone makes text big**
![image](https://user-images.githubusercontent.com/26465327/80214918-8cafb180-8633-11ea-8e2a-96ae35673621.png)

**!! Only makes text bolt**
![image](https://user-images.githubusercontent.com/26465327/80214926-8f120b80-8633-11ea-917d-acc9dbfd0878.png)

**Colour is based on job**
![image](https://user-images.githubusercontent.com/26465327/80214937-933e2900-8633-11ea-8e4a-27c9dff58fba.png)
![image](https://user-images.githubusercontent.com/26465327/80214939-9507ec80-8633-11ea-9c15-11fd36925d39.png)
![image](https://user-images.githubusercontent.com/26465327/80214944-96d1b000-8633-11ea-887b-3624ab748430.png)
![image](https://user-images.githubusercontent.com/26465327/80214949-9802dd00-8633-11ea-935a-a1cdb2f26f15.png)
![image](https://user-images.githubusercontent.com/26465327/80214952-99340a00-8633-11ea-8c98-523537a14f55.png)
![image](https://user-images.githubusercontent.com/26465327/80214958-9a653700-8633-11ea-8f72-351a82b7fa1c.png)
![image](https://user-images.githubusercontent.com/26465327/80214962-9c2efa80-8633-11ea-8633-eb61eb254cba.png)
![image](https://user-images.githubusercontent.com/26465327/80214966-9d602780-8633-11ea-9de9-511c6700389a.png)
![image](https://user-images.githubusercontent.com/26465327/80214970-9e915480-8633-11ea-97e0-45e9d77498bf.png)
![image](https://user-images.githubusercontent.com/26465327/80214973-a05b1800-8633-11ea-81ac-e03c34445bbf.png)
![image](https://user-images.githubusercontent.com/26465327/80214977-a0f3ae80-8633-11ea-8f9b-d53bd7553132.png)
![image](https://user-images.githubusercontent.com/26465327/80214981-a2bd7200-8633-11ea-81ac-12680acbe744.png)
![image](https://user-images.githubusercontent.com/26465327/80214984-a3ee9f00-8633-11ea-8de8-4ab2e2967c36.png)
![image](https://user-images.githubusercontent.com/26465327/80214990-a5b86280-8633-11ea-9c0a-d9adf985107c.png)
![image](https://user-images.githubusercontent.com/26465327/80214994-a6e98f80-8633-11ea-96cb-8fd909cb1ec5.png)
![image](https://user-images.githubusercontent.com/26465327/80215000-a81abc80-8633-11ea-94b0-c701502312d9.png)
![image](https://user-images.githubusercontent.com/26465327/80215004-a94be980-8633-11ea-82f3-4cecb4bdc5ac.png)
![image](https://user-images.githubusercontent.com/26465327/80215012-ab15ad00-8633-11ea-9083-9f13386353f9.png)
![image](https://user-images.githubusercontent.com/26465327/80215018-acdf7080-8633-11ea-90b5-2afdf17e2720.png)
![image](https://user-images.githubusercontent.com/26465327/80215020-af41ca80-8633-11ea-9733-6ab872bd1428.png)
![image](https://user-images.githubusercontent.com/26465327/80215022-b072f780-8633-11ea-8791-a1990fbe18e4.png)
![image](https://user-images.githubusercontent.com/26465327/80215025-b1a42480-8633-11ea-8906-f3529577dc5d.png)
![image](https://user-images.githubusercontent.com/26465327/80215033-b4067e80-8633-11ea-9d03-07a14bb2bc61.png)
![image](https://user-images.githubusercontent.com/26465327/80215037-b537ab80-8633-11ea-9515-f9e03fedf00f.png)
![image](https://user-images.githubusercontent.com/26465327/80215040-b668d880-8633-11ea-8486-3a9e6f864bc3.png)
![image](https://user-images.githubusercontent.com/26465327/80215044-b79a0580-8633-11ea-935d-8551d9aa3453.png)
![image](https://user-images.githubusercontent.com/26465327/80215046-b8cb3280-8633-11ea-9eca-2ea029dc4c68.png)
![image](https://user-images.githubusercontent.com/26465327/80215047-b9fc5f80-8633-11ea-98a3-96aca8db54a4.png)
![image](https://user-images.githubusercontent.com/26465327/80215050-bbc62300-8633-11ea-84cb-8cd5d30188fd.png)
![image](https://user-images.githubusercontent.com/26465327/80215059-c1236d80-8633-11ea-8c46-5deed53ab15e.png)
![image](https://user-images.githubusercontent.com/26465327/80215062-c2549a80-8633-11ea-9f7b-acd3e4f0e41b.png)
![image](https://user-images.githubusercontent.com/26465327/80215066-c385c780-8633-11ea-8a5d-0e38a20b00ed.png)
![image](https://user-images.githubusercontent.com/26465327/80215073-c54f8b00-8633-11ea-9f00-4653efed3a56.png)
![image](https://user-images.githubusercontent.com/26465327/80215077-c680b800-8633-11ea-988b-08a20a91bdf7.png)
![image](https://user-images.githubusercontent.com/26465327/80215053-bcf75000-8633-11ea-8b51-ed0fdfc8f76a.png)
</p>
</details>

## Why It's Good For The Game

Makes the new cool chat have some nicer looking formatting and colours, which is good.

Non-humans have random colour still, anyone without an ID will be grey and unknown/custom jobs will show up as an orange colour.

## Changelog
:cl:
tweak: overhead chat now based on job name.
tweak: certain simplemobs have custom colours too
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

PowerfulBacon committed:
Changed burgerchat colours to be based on job

ghost commented:
WHY IS EVERY CHARACTER IN THOSE IMAGES A FELINID

PowerfulBacon commented:
> 
> 
> WHY IS EVERY CHARACTER IN THOSE IMAGES A FELINID

because its the same character with job changes lol

PowerfulBacon commented:
ah fuck I gotta do some git magic to fix the damn changelog

PowerfulBacon committed:
Makes whisper smaller again

PowerfulBacon committed:
reverts 5px text because 6 is better

PowerfulBacon commented:
![image](https://user-images.githubusercontent.com/26465327/80243841-d9a97d00-865f-11ea-9c07-aff5225e9896.png)

CthulhuOnIce reviewed (changes requested):
I said this already in discord, but I'm also going to say it here for record-keeping and organization:

- If you can find a better way to attach pretty colors to jobs please use that, as coders will probably forget or won't bother to update the `job_colors_pastel` list

- Please update the chat layer so that it's above everything **except** lighting; as it randomly appears behind overlays

PowerfulBacon commented:
> 
> 
> I said this already in discord, but I'm also going to say it here for record-keeping and organization:
> 
>     * If you can find a better way to attach pretty colors to jobs please use that, as coders will probably forget or won't bother to update the `job_colors_pastel` list
> 
>     * Please update the chat layer so that it's above everything **except** lighting; as it randomly appears behind overlays

Will work on this tomorrow, got the lighting layer thing done, I'll figure something out for the colours to make it easier for coders.

CthulhuOnIce commented:
epic, sounds gucci, which layer did you use?

PowerfulBacon commented:
> 
> 
> epic, sounds gucci, which layer did you use?

Made a new one #define CHAT_TEXT_LAYER 5.2 so it isn't on the same layer as anything and will always be consistant (above ripple, just above fly but also below ghost layers, maybe should change that)

qwertyquerty commented:
Why should it not be above lighting?

PowerfulBacon commented:
> 
> 
> Why should it not be above lighting?

As Kev said on discord, would look weird being in pitch black with pastel text, and it is less spookyness in dark with big text

PowerfulBacon commented on code/modules/mob/living/burgerchat.dm:
```diff
@@ -1,40 +1,93 @@
+	var/mob/living/carbon/human/target_as_human = target
+	if(istype(target_as_human))
+		if(target_as_human.wear_id?.GetID())
+			var/datum/job/wearer_job = target_as_human.wear_id.GetJobName()
```
This variable shouldn't work, but I am currently refactoring this to be easier to add new jobs

WhyIsCaeciliusTaken commented:
Won't this let you identify unknowns and nukeops etc? What about changelings?

PowerfulBacon commented:
> 
> 
> Won't this let you identify unknowns and nukeops etc? What about changelings?

If you have an ID card in your ID slot it will use the job name of that ID card. Forged IDs with job name Assistant will take on the assistant colour.
If you do not have an ID it will be grey, if the job is custom / not a standard job it will be orange.

ike709 commented:
If you could grab the URL regex from here and include it that'd be great: https://github.com/tgstation/tgstation/pull/50608/files

```
// Get rid of any URL schemes that might cause BYOND to automatically wrap something in an anchor tag
var/static/regex/url_scheme = new(@"[A-Za-z][A-Za-z0-9+-\.]*:\/\/", "g")
text = replacetext(text, url_scheme, "")
```

ike709 commented on code/modules/mob/living/burgerchat.dm:
```diff
@@ -1,40 +1,93 @@
 /mob/living
 	var/list/stored_chat_text = list()
 
+/proc/get_job_colour(job_title)
```
Seems like unnecessary proc overhead.

PowerfulBacon replied:
Proc phased out anyway

bobbah commented:
> If you could grab the URL regex from here and include it that'd be great: https://github.com/tgstation/tgstation/pull/50608/files
> 
> ```
> // Get rid of any URL schemes that might cause BYOND to automatically wrap something in an anchor tag
> var/static/regex/url_scheme = new(@"[A-Za-z][A-Za-z0-9+-\.]*:\/\/", "g")
> text = replacetext(text, url_scheme, "")
> ```

:flushed:

bobbah commented on code/modules/mob/living/burgerchat.dm:
```diff
@@ -1,40 +1,93 @@
+	"Brig Physician" = 	"#B364B3",
+	"Prisoner" = 		"#d38a5c",
+	"CentCom" = 		"#90FD6D",
+	"Unknown"=			COLOR_JOB_UNKOWN,
```
COLOR_JOB_UNKOWN

typo?

PowerfulBacon replied:
well spotted

PowerfulBacon commented:
Can you change the layer of map text? Changing the layer of the object does not affect the map text.

PowerfulBacon committed:
Updates the thing so its easier for people to not forget to add new chat colours

PowerfulBacon committed:
Regex thingy

PowerfulBacon commented:
@CthulhuOnIce Cannot seem to change the layer, changing the layer of the image doesn't change the layer of the maptext. But I have updated the job thing and added the regex thing

PowerfulBacon committed:
Removed CHAT_TEXT_LAYER since it didn't do anything

PowerfulBacon committed:
job unknown spelling fix

CthulhuOnIce reviewed (approved):
good

CthulhuOnIce commented:
did the command line shit but it didnt close this automatically, so

CthulhuOnIce commented:
but this has been effectively merged and had conflicts removed

ike709 commented:
What a stupid way of merging a PR.

ike709 commented on code/modules/mob/living/burgerchat.dm:
```diff
@@ -1,40 +1,69 @@
+		else
+			text_color = COLOR_PERSON_UNKNOWN
+	else
+		text_color = pick(chatOverhead_colors)
```
Remove the list, it's redundant now and we can just use `COLOR_JOB_UNKNOWN`.

PowerfulBacon replied:
I like the mobs like Ian having nicer say colours than hostile mobs, so I'm thinking of adding a var on mob that lets you choose the chat colour in case the mob is not of type /mob/living/carbon/human

PowerfulBacon committed:
Adds in mobsay_color var so certain mobs can have different say colours

PowerfulBacon committed:
Merge branch 'master' into runescape-chat-based-on-job

PowerfulBacon commented:
Testing the merge conflicts, because of my history with resolving merge conflicts does not give me faith I did it right

PowerfulBacon committed:
Removes whispering from overhead chat

PowerfulBacon committed:
Travis rebuild attempt

CthulhuOnIce commented:
alright, it built, is this ready?

PowerfulBacon commented:
> alright, it built, is this ready?

Should be yes
